src/messaging/ExchangeMgr.cpp

```diff
@@ -104,6 +104,14 @@ CHIP_ERROR ExchangeManager::Shutdown()
     mMessageCounterSyncMgr.Shutdown();
     mReliableMessageMgr.Shutdown();
 
+    for (auto & ec : mContextPool)
+    {
+        if (ec.GetReferenceCount() > 0)
+        {
+            ec.Abort();
```

Comment:
So, a few issues here:

1. If there is more than one outstanding reference to `ec`, this will just do a single Release and still not actually call `Free` on the `ExchangeContext`.
2. If there is only one outstanding reference, whenever the holder of that reference goes to release it, it will be doing things with memory it no longer owns, right?

Fundamentally, this seems like it's working around some sort of failure to follow an API contract somewhere is.  Should we not fix those failures?  And then perhaps we should take more drastic action (crashing?) if the API contract is violated in ways that we cannot recover from?

Reply:
Yes, there are problems in the PR. Here I want to force free the exchange context, and it will trigger `OnExchangeClosing` to notify all application, so they get a chance to release resources associated with the exchange. I'll update the PR to do it correctly.

I also have another PR, which totally changes the schematic meaning of `ExchangeContext::Close` and `ExchangeContext::Abort`.

`ExchangeContext::Close`: gracefully close the exchange
`ExchangeContext::Abort`: force fail the exchange

Both of them will not modify the reference counter, unique_ptr like handle pointing to the exchange context will be used to trace the usage count.